go/vt/vttablet/tabletmanager/vreplication/vreplicator.go

```diff
@@ -257,47 +259,56 @@ func (vr *vreplicator) buildPkInfoMap(ctx context.Context) (map[string][]*Primar
 		} else {
 			pks = td.Columns
 		}
-		var pkInfos []*PrimaryKeyInfo
-		for _, pk := range pks {
+		var colInfo []*ColumnInfo
+		for _, row := range qr.Rows {
 			charSet := ""
 			collation := ""
+			columnName := ""
+			isPK := false
+			isGenerated := false
 			var dataType, columnType string
-			for _, row := range qr.Rows {
-				columnName := row[2].ToString()
-				if strings.EqualFold(columnName, pk) {
-					var currentField *querypb.Field
-					for _, field := range td.Fields {
-						if field.Name == pk {
-							currentField = field
-							break
-						}
-					}
-					if currentField == nil {
-						continue
-					}
-					dataType = row[3].ToString()
-					columnType = row[4].ToString()
-					if sqltypes.IsText(currentField.Type) {
-						charSet = row[0].ToString()
-						collation = row[1].ToString()
-					}
+			columnName = row[2].ToString()
+			var currentField *querypb.Field
+			for _, field := range td.Fields {
+				if field.Name == columnName {
+					currentField = field
 					break
 				}
 			}
+			if currentField == nil {
+				continue
+			}
+			dataType = row[3].ToString()
+			columnType = row[4].ToString()
+			if sqltypes.IsText(currentField.Type) {
+				charSet = row[0].ToString()
+				collation = row[1].ToString()
+			}
 			if dataType == "" || columnType == "" {
-				return nil, fmt.Errorf("no dataType/columnType found in information_schema.columns for table %s, column %s", td.Name, pk)
+				return nil, fmt.Errorf("no dataType/columnType found in information_schema.columns for table %s, column %s", td.Name, columnName)
+			}
+			for _, pk := range pks {
+				if columnName == pk {
+					isPK = true
+				}
+			}
+			extra := row[5].ToString()
+			if strings.Contains(strings.ToLower(extra), "generated") {
```

Comment:
also check for `"virtual"` because it could be `"virtual stored"`. See https://github.com/vitessio/vitess/pull/8043/files